pkg/server/endpoints/middleware.go

```diff
@@ -0,0 +1,189 @@
+	return middleware.EntryFetcherFunc(func(ctx context.Context, id spiffeid.ID) ([]*types.Entry, error) {
+		resp, err := ds.ListRegistrationEntries(ctx, &datastore.ListRegistrationEntriesRequest{
+			BySpiffeId: &wrappers.StringValue{
+				Value: id.String(),
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+		return api.RegistrationEntriesToProto(resp.Entries)
+	})
+}
+
+func AuthorizedEntryFetcher(ds datastore.DataStore) api.AuthorizedEntryFetcher {
+	return api.AuthorizedEntryFetcherFunc(func(ctx context.Context, agentID spiffeid.ID) ([]*types.Entry, error) {
+		entries, err := regentryutil.FetchRegistrationEntries(ctx, ds, agentID.String())
+		if err != nil {
+			return nil, err
+		}
+		return api.RegistrationEntriesToProto(entries)
+	})
+}
+
+func AgentAuthorizer(log logrus.FieldLogger, ds datastore.DataStore) middleware.AgentAuthorizer {
+	return middleware.AgentAuthorizerFunc(func(ctx context.Context, agentID spiffeid.ID, agentSVID *x509.Certificate) error {
+		id := agentID.String()
+		log := log.WithField(telemetry.AgentID, id)
+
+		resp, err := ds.FetchAttestedNode(ctx, &datastore.FetchAttestedNodeRequest{
+			SpiffeId: id,
+		})
+		switch {
+		case err != nil:
+			log.WithError(err).Error("Unable to look up agent information")
+			return status.Errorf(codes.Internal, "unable to look up agent information")
+		case resp.Node == nil:
+			log.Error("Agent is not attested")
+			return status.Errorf(codes.PermissionDenied, "agent %q is not attested", id)
+		case resp.Node.CertSerialNumber == "":
+			log.Error("Agent is banned")
+			return status.Errorf(codes.PermissionDenied, "agent %q is banned", id)
+		case resp.Node.CertSerialNumber != agentSVID.SerialNumber.String():
+			log.WithFields(logrus.Fields{
+				telemetry.SVIDSerialNumber: agentSVID.SerialNumber.String(),
+				telemetry.SerialNumber:     resp.Node.CertSerialNumber,
+			}).Error("Agent SVID is stale")
+			return status.Errorf(codes.PermissionDenied, "agent %q expected to have serial number %q; has %q", id, resp.Node.CertSerialNumber, agentSVID.SerialNumber)
+		default:
+			return nil
+		}
+	})
+}
+
+func RateLimits() map[string]api.RateLimiter {
+	noLimit := middleware.NoLimit()
+	attestLimit := middleware.PerIPLimit(node_pb.AttestLimit)
+	csrLimit := middleware.PerIPLimit(node_pb.CSRLimit)
+	jsrLimit := middleware.PerIPLimit(node_pb.JSRLimit)
+	pushJWTKeyLimit := middleware.PerIPLimit(node_pb.PushJWTKeyLimit)
+
+	return map[string]api.RateLimiter{
+		"/spire.api.server.svid.v1.SVID/MintX509SVID":                   noLimit,
+		"/spire.api.server.svid.v1.SVID/MintJWTSVID":                    noLimit,
+		"/spire.api.server.svid.v1.SVID/BatchNewX509SVID":               csrLimit,
+		"/spire.api.server.svid.v1.SVID/NewJWTSVID":                     jsrLimit,
+		"/spire.api.server.svid.v1.SVID/NewDownstreamX509CA":            csrLimit,
+		"/spire.api.server.bundle.v1.Bundle/GetBundle":                  noLimit,
```

Comment:
I know that this is per current design/implementation, but I'm wondering if we should be more strict in the limits. Particularly in this call that is not authenticated (although it shouldn't have a high cost) and it could be easily abused.

Reply:
I think this vector for abuse already exists, since TLS handshakes cause a bundle lookup. I know @APTy has been working on a solution to mitigate that though.

What sort of limit did you have in mind?

Reply:
I was just thinking if it would make sense to have a per-ip limit here. Probably a decision out of scope for this PR.